packages/fondue/src/components/RichTextEditor/BlurObserver.ts

```diff
@@ -27,35 +30,58 @@ export const BlurObserver = ({
 
     useEffect(() => {
+        if (!isBlurObserverActive) {
+            return;
+        }
+
         const handleInteractWithOutside = (event: FocusEvent | MouseEvent) => {
             const richTextEditorSelector = `[data-editor-id='${editorRef.id}']`;
             const toolbarSelector = "[data-selector='toolbar-floating']";
             const modalSelector = '[data-is-underlay]';
-            const richTextEditorWrapper = document.querySelector<HTMLDivElement>(richTextEditorSelector);
-            const toolbarWrapper = document.querySelector<HTMLDivElement>(toolbarSelector);
-            const modalWrapper = document.querySelector<HTMLDivElement>(modalSelector);
+
             const targetElement = event.target as HTMLElement;
 
-            const isTargetInsideModal = modalWrapper?.contains(targetElement);
-            const isTargetInsideRichTextEditor = richTextEditorWrapper?.contains(targetElement);
-            const isTargetInsideRichTextEditorToolbar = toolbarWrapper?.contains(targetElement);
+            const isInsidePortalledModal = targetElement.closest(modalSelector);
+            const isInsideNonPortalledModal = targetElement.closest(FLOATING_MODAL_SELECTOR);
+            const isInsideRichTextEditor = targetElement.closest(richTextEditorSelector);
+            const isInsideRichTextEditorToolbar = targetElement.closest(toolbarSelector);
 
-            if (isTargetInsideRichTextEditorToolbar) {
+            if ((isInsideRichTextEditor && !isInsideNonPortalledModal) || isInsideRichTextEditorToolbar) {
                 collapseFloatingModals();
-            }
-
-            if (!isTargetInsideRichTextEditor && !isTargetInsideRichTextEditorToolbar && !isTargetInsideModal) {
+            } else if (!isInsideRichTextEditor && !isInsideRichTextEditorToolbar && !isInsidePortalledModal) {
                 collapseEverything();
+                setIsBlurObserverActive(false);
             }
         };
 
         document.addEventListener('pointerdown', handleInteractWithOutside);
         document.addEventListener('focusin', handleInteractWithOutside);
+
         return () => {
             document.removeEventListener('focusin', handleInteractWithOutside);
             document.removeEventListener('pointerdown', handleInteractWithOutside);
         };
-    }, [collapseEverything, collapseFloatingModals, editorRef.id]);
+    }, [collapseEverything, collapseFloatingModals, editorRef.id, isBlurObserverActive]);
+
+    /* Only activate blur observer when the user has focused the editor
+     to limit the number of active document event listeners */
+    useEffect(() => {
+        const richTextEditor = document.querySelector(`[data-editor-id='${editorRef.id}']`);
+
+        if (!richTextEditor) {
+            return;
+        }
+
+        const setupBlurObserver = () => {
+            setIsBlurObserverActive(true);
+        };
+
+        richTextEditor.addEventListener('focusin', setupBlurObserver);
```

Comment:
Shall we unsubscribe on focusout? Otherwise click in => click out => click in will have 2 listeners for the same thing, right?

Reply:
click out should always trigger the unsubscribe from the hook above

Reply:
adding the focusout handler means repeating a lot of the conditional checks that are already prevent in the document event listener (check if its in a modal or the toolbar etc). It seems like it would make sense for it to unsubscribe itself

Reply:
yeah my bad